CRAB-33767: Incorporate Snapshots and AMIs in NagBot

alizbay94 opened the pull request:
This PR introduces changes that allows NagBot to process AMIs and snapshots when issuing the 'execute' and 'notify' commands. 

**Pricing for Snapshots** was determined using the average price of storing _standard_ Snapshots in us-west-1, us-west-2, us-east-1, and us-east-2. I did not include logic to estimate the price of _archived_ snapshots since Seeq only uses standard snapshots to-date.

**Pricing for AMIs** was determined by taking the sum total price of the snapshots comprising the AMI. Only pricing for 
_EBS-Backed_ AMI's was implemented since Seeq does not use _instance-backed_ AMIs to-date.

**Terminating AMIs [considerations](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/deregister-ami.html#clean-up-ebs-ami)**:
- When you deregister an EBS-backed AMI, it doesn't affect the snapshot(s) that were created for the volume(s) of the instance during the AMI creation process. You'll continue to incur storage costs for the snapshots. I did not implement logic to explicitly delete the snapshots comprising EBS-backed AMIs since ideally those snapshots would have the same terminate date as the AMI upon creation. NagBot processes AMIs before snapshots, so the AMI would be deregistered then the snapshots comprising it would be deleted soon-after if they have the same terminate date. I can still implement the logic to also delete the snapshots when the AMI is deregistered if it sounds desirable, the snapshot ids are privided by the `boto3.resource.Image()` so it should only be a matter of calling `terminate_resource` on those snapshots.
**Note** After further discussion, functionality has been implemented in this PR so that any Snapshots making up an EBS-backed AMI are deleted when NagBot deregisters the AMI. Furthermore, pricing for Snapshots belonging to EBS-Backed AMIs are not included in the monthly Snapshot pricing, but they are instead included in the monthly AMI pricing.

**Terminating Snapshot [considerations](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-deleting-snapshot.html#ebs-delete-snapshot-considerations)**:
- You can't delete a snapshot of the root device of an EBS volume used by a registered AMI. You must first deregister the AMI before you can delete the snapshot. For this reason, AMI's are processed before snapshots when NagBot runs so that the AMI is deregistered before the snapshots comprising it are deleted. I didn't implement logic to deregister the AMI which the snapshot is a part of in this scenario, since the AMI and the snapshots it is composed of should have the same terminate date if we implement an AWS Lambda function in the future that automatically tags resources. Since AMIs are processed by NagBot first, the AMI would be deregistered if its terminate date comes up, then the snapshots would be able to be deleted - the snapshots would be deleted when NagBot processes our organizations snapshots.
- You can't delete a snapshot that is managed by the AWS Backup service using Amazon EC2. Instead, you must use AWS Backup to delete the corresponding recovery points in the backup vault - I did not implement logic for this scenario, but I imagine I could if it is desirable.

**Pytests have been written** for the `Snapshot` and `Ami` classes. 3/5 tests written for each class are identical to the tests written for the `Volume` class, with the exception of 2 tests involving terminating resources. Since AMIs and snapshots are deleted differently with the python `boto3` library than how EBS volumes and EC2 instances are deleted, I set the return value the aws `Snapshot()`and `Image()` classes to a MagicMock() - eventually it is asserted that the MagicMock() called `delete()` and `deregister()` for snapshots and ami respectively during `terminate_resource`.

**`README.md` has been updated** with instructions on how to update NagBot's version. Mentioning of volumes, AMIs, and snapshots has also been included in the README. A new slack app for testing NagBot has been created and added to the #bot-testing channel in slack., so the`README.md` was also updated with how to test NagBot during development, and where to find the `SLACK_BOT_TOKEN` for the NagBotTest slack app on Keeper.

**Some refactoring was done** as part of the changes in this PR so that `format()` strings were converted to f strings for readability.

---

Future Improvements: I noticed that our organization's snapshots and AMIs are often untagged, implementing a method of auto-tagging resources in the future (potentially using an AWS Lambda) would be beneficial to ensure those resources get tagged and deleted when they are no longer needed.

alizbay94 committed:
Create files for Ami and Snapshot classes, and create dataclasses for each

alizbay94 committed:
Implement execute function for Snapshots and refactor strings to use f strings instead of .format()

alizbay94 committed:
Merge branch 'master' into task/azb/support-ebs-snapshot-and-amis-CRAB-33767

alizbay94 committed:
Include is_stoppable function for Snapshot class

alizbay94 committed:
Estimate monthly cost for AMI and terminate AMIs

alizbay94 committed:
Add pytests for AMIs

alizbay94 committed:
Add tests for Snapshots

alizbay94 committed:
Implement logic for processing AMIs with ephemeral volumes

alizbay94 committed:
Refactor comments and remove TODOs that are no longer relevant

alizbay94 committed:
Use MagicMock() to mock the return value of client and resorce for ami and snapshot termination pytests

alizbay94 committed:
Use VolumeSize to reference size of snapshot in snapshot_dictionary

alizbay94 committed:
Update README.md

alizbay94 committed:
Include instructions for testing development changes to NagBot

alizbay94 committed:
Process AMIs before snapshots

CodyRay commented on README.md:
```diff
@@ -22,6 +25,13 @@ Here's what a Nagbot notification looks like in Slack:
+The NagBotTest app has been registered to slack, and it is set up to send NagBot output to slack. The default 
+testing channel for NagBot is #bot-testing. 
+
+The SLACK_BOT_TOKEN for NagBotTest is found on Keeper [here](https://keepersecurity.com/vault/#detail/AUj3FYXz44uON4CVQSTKMQ).
```
:+1: Remember that this repository is public, so consider prefacing this to be for Seeq employees in case anyone stumbles upon nagbot

alizbay94 replied:
done

CodyRay commented on app/ami.py:
```diff
@@ -0,0 +1,158 @@
+import boto3
+
+from datetime import datetime
+TODAY = datetime.today()
```
(sanity check) are any changes needed here from #11?

alizbay94 replied:
I don't think any changes would be necessary since that PR was only affecting `resource.py`, @sejal-raval does this sound right?

sejal-raval replied:
That's a good point. https://github.com/seeq12/nagbot/pull/11 changed that all generic methods in `resource.py` reads time in Pacific TimeZone. I think it should be changed for the individual resources e.g. instances, volumes etc. for concurrency. I can open another PR for that.

CodyRay commented on app/ami.py:
```diff
@@ -0,0 +1,158 @@
+                self.region_name,
+                self.resource_type,
+                self.operating_system,
+                self.iops,
```
Is storage type (gp2, gp3, for example) also available?

alizbay94 replied:
storage and iops are only available on some amis, but I just implemented a function that assigns iops and volume_type if it exists in the block_device_mappings, otherwise iops and volume_type will just be None.

CodyRay commented on app/ami.py:
```diff
@@ -0,0 +1,158 @@
+    def build_model(region_name: str, resource_dict: dict):
+        tags = resource.make_tags_dict(resource_dict.get('Tags', []))
+
+        state = resource_dict['State']
```
What kind of states does an AMI have?

alizbay94 replied:
I believe the only states are available and pending

CodyRay commented on app/instance.py:
```diff
@@ -24,7 +23,7 @@ class Instance(Resource):
     @staticmethod
     def to_string():
-        return 'instance', 'running'
+        return 'instance', 'running'  # TODO: why do we assume this is running?
```
Good question, I thought this was only used for building messages?

alizbay94 replied:
I think it is only for building messages, so I left it as is even though an AMI doesn't necessarily run.

CodyRay commented on app/nagbot.py:
```diff
@@ -33,63 +35,59 @@
             ec2_type, ec2_state = resource_type.to_string()
             resources = resource_type.list_resources()
-            num_running_resources = sum(1 for r in resources if r.state == 'running' or r.state == 'available')
+            num_running_resources = sum(1 for r in resources if r.state in ['running', 'available', 'completed'])
```
(refactoring opportunity) resources could have a `is_running` method that handles this to avoid the state leaking out like this

alizbay94 replied:
Done, I added a method called `is_active` to each of the resources

CodyRay commented on app/nagbot.py:
```diff
@@ -33,63 +35,59 @@
             if len(resources_to_terminate) > 0:
-                summary_msg += 'The following %d _stopped_ {}s are due to be *TERMINATED*, ' \
-                                'based on the "Terminate after" tag:\n'.format(ec2_type) \
-                                % len(resources_to_terminate)
```
Oof, this one had both `.format` and `%` based formatting - nice to see strings cleaned up :+1:

CodyRay commented on app/nagbot.py:
```diff
@@ -33,63 +35,59 @@
             self.notify_internal(channel, dryrun)
         except Exception as e:
-            sqslack.send_message(channel, "Nagbot failed to run the 'notify' command: " + str(e))
+            sqslack.send_message(channel, f"Nagbot failed to run the 'notify' command: {str(e)}")
```
(nit) `str()` is redundant here now

CodyRay commented on app/nagbot.py:
```diff
@@ -108,37 +106,36 @@ def execute_internal(channel, dryrun):
-                sqslack.send_message(channel, 'No {}s were stopped today.'.format(ec2_type))
+                sqslack.send_message(channel, f'No {ec2_type}s were terminated today.')
+
+            if resource_type == "Instance":
```
Why is this needed? Shouldn't snapshots and amis use the same method as volumes to skip this part?

alizbay94 replied:
I implemented this because the output on slack includes "No volumes are due to be stopped at this time" and since Volumes, Snapshots, and AMIs don't stop I thought it would look cleaner on slack to only display _stopped_ Instances.

CodyRay replied:
That seems like a good idea. Maybe there is a way to define if a resource type is "stopable" generically?

alizbay94 replied:
I see, so it would be

`if resource_type.can_be_stopped()`

rather than 

`if resource_type == "Instance"`

I can implement that

alizbay94 replied:
done

CodyRay commented on app/resource.py:
```diff
@@ -74,6 +74,28 @@ def estimate_monthly_ebs_storage_price(region_name: str, instance_id: str, volum
 
 
+# Estimated monthly costs were formulated by taking the average monthly costs of N. California and Oregon
+def estimate_monthly_snapshot_price(type: str, size: float) -> float:
```
(refactoring opportunity) IMO many (all?) of the methods in this file could belong in their corresponding file (instance.py, etc)

alizbay94 replied:
I agree, i'll make those changes. It just doesn't feel right putting those methods in the resource file 😄

alizbay94 replied:
Done, moved the estimate monthly cost functions from resource.py to their respective files

CodyRay commented on app/resource.py:
```diff
@@ -74,6 +74,28 @@ def estimate_monthly_ebs_storage_price(region_name: str, instance_id: str, volum
+def estimate_monthly_ami_price(type: str, block_device_mappings: list, ami: str) -> float:
+    total_cost = 0
+    # Logic is only implemented for ebs-backed AMIs since Seeq does not use instance-backed AMIs
+    if type == 'ebs':
```
I'd vote to log a warning if we get a type that isn't ebs

alizbay94 replied:
Done I used a print statement with details of the AMI if it is not ebs-backed

CodyRay commented on app/ami.py:
```diff
@@ -0,0 +1,158 @@
+
+    # Check if an ami is deletable/terminatable
+    def is_terminatable(self, today_date):
+        state = 'available'
```
Is it possible to determine if the snapshot is associated with an AMI or AWS backup here to prevent us from trying to delete the snapshot?

alizbay94 replied:
Yes, that information is included in the "Description" of the dictionary returned by the aws call - I just added two additional attributes in the Snapshot object, `is_ami_snapshot` and `is_aws_backup` snapshot. If the snapshot is either it will not get deleted, and I adjusted the pytests to test for that scenario.

CodyRay commented on app/ami.py:
```diff
@@ -0,0 +1,158 @@
+        return Resource.generic_url_from_id(region_name, resource_id, resource_type)
+
+    # Include ami in monthly price calculation if available
+    def included_in_monthly_price(self):
```
I think currently the snapshots associated with amis will get double counted (counted for snapshot total and ami total). I'd vote to count the associated ones with the AMI total. Which I think you could do here by returning false if the snapshot is associated with an ami

alizbay94 replied:
Done, I made it so that if it `is_ami_snapshot` its not included in the monthly price

CodyRay reviewed (approved):
Looking good, mostly learning questions from me

> I did not implement logic to explicitly delete the snapshots comprising EBS-backed AMIs since ideally those snapshots would have the same terminate date as the AMI upon creation.

I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)

alizbay94 commented:
> I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)

Sounds good i'll implement that so they are all deleted as one happy family.

alizbay94 committed:
Update README so that devs know keeper link is for seeq employees
:

alizbay94 committed:
Implement function to extract ami and iops from block_device_mappings

alizbay94 committed:
Include volume_type in ami class in pytests

alizbay94 committed:
Merge branch 'master' into task/azb/support-ebs-snapshot-and-amis-CRAB-33767

alizbay94 committed:
Log error if ami type is not ebs-backed

alizbay94 committed:
Determine if snapshot is aws backup snapshot or ami snapshot, and don't include ami snapshot in monthly price

alizbay94 committed:
Do not delete aws backup or ami snapshots, and implement pytests for both circumstances

alizbay94 committed:
Remove redundant str()

alizbay94 committed:
Create method for checking if resources are active

alizbay94 committed:
Move monthly estimated cost methods from resource.py to their respective files

CodyRay commented on app/ami.py:
```diff
@@ -143,6 +144,10 @@ def is_safe_to_terminate(self, today_date):
 
+    # Check if a instance is active
+    def is_active(self):
+        return True if self.state == 'available' else False
```
Can be simplified to 

```suggestion
        return self.state == 'available'
```

CodyRay commented on app/ami.py:
```diff
@@ -176,3 +181,22 @@ def get_ami_iops_and_volume_type(block_device_mappings):
+                total_cost += estimate_monthly_snapshot_price(snapshot_type, snapshot_size)
+    else:
+        print(f"WARNING: {ami_name} is a {ami_type} type AMI with the following block_device_mappings: "
+              "{block_device_mappings}")
```
Looks like an `f` was lost
```suggestion
              f"{block_device_mappings}")
```

CodyRay commented on app/snapshot.py:
```diff
@@ -77,30 +83,39 @@ def build_model(region_name: str, resource_dict: dict):
+        is_ami_snapshot = False
+        if "AWS Backup service" in resource_dict['Description']:
+            is_aws_backup_snapshot = True
+        elif "Copied for DestinationAmi" in resource_dict['Description']:
```
This might not be the most rebust method for Amis in particular because if someone deletes an AMI manually and doesn't delete the snapshot, then nagbot won't ever clean it up.

alizbay94 replied:
Good point. I could extract the AMI's ID from the description of the snapshot, check if it exists with boto3, if it doesn't then the snapshot could be considered terminatable / part of the overall cost.

alizbay94 replied:
Implemented logic for the following:
If snapshot is an ami snapshot, we use a regex to match the first occurrence of an ami id since the first occurrence refers to the 'Destination AMI' (what we want) instead of the 'Source AMI'.
That ami id is checked against all the orgs amis to see if it is registered - if it is not registered the snapshot is not considered an ami snapshot and it will be treated like the rest of its bretheren able to be deleted and included in the total snapshot cost. (After running this looks like we have a good amount of ami snapshots whose amis have been deregistered so we should save a nice chunk of change by deleting them!)
One note - I used a global variable for the orgs amis so that the API call would only be made rather than once for every ami snapshot since the call is costly.

alizbay94 committed:
Update app/ami.py with f string

Co-authored-by: Cody Ray Hoeft <cody.ray.hoeft@seeq.com>

alizbay94 committed:
Simplify return value for is_active in app/ami.py

Co-authored-by: Cody Ray Hoeft <cody.ray.hoeft@seeq.com>

alizbay94 committed:
Determine if snapshot is ami snapshot - if destination snapshot is deregistered do not count as ami snapshot

alizbay94 committed:
Implement generic can_be_stopped() method and refactor slack message to have cleaner appearance

alizbay94 committed:
Refactor comments for can_be_stopped() for each resource

CodyRay commented on app/ami.py:
```diff
@@ -198,5 +204,15 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
+        ec2_client = boto3.client('ec2')
+        ORG_AMIS = ec2_client.describe_images(Owners=['self'])
+    # Iterate through each image owned by organization to check if ami id exists, if it does it is registered
+    return True if [i['ImageId'] for i in ORG_AMIS['Images'] if i['ImageId'] == ami_id] else False
```
Instead of loading a list of images, I think you could request just `ami_id` to load a single image. If it succeeds, then it must be registered. If boto3 fails to load it then it probably doesn't exist. I think that would be efficient enough that a global variable wouldn't be needed

alizbay94 replied:
Done.

I ended up using 2 exceptions for the try/except  attempting to find a single AMI with boto3 because typically, if an ami does not exist, `ec2.Image(ami_id).name` throws a `botocore.exceptions.ClientError`, but I encountered a case where an `AttributeError` is thrown that reads: `AttributeError: 'NoneType' object has no attribute 'get'`

After doing a google search, I discovered a bug with an open ticket [here](https://github.com/boto/boto3/issues/925) on the boto3 project’s GitHub page describing how None type is returned for deregistered images, so I included the `AttributeError` as well to cover those cases.

alizbay94 committed:
Check if AMI exists using a single call rather than getting list of org's amis and checking

CodyRay commented on app/ami.py:
```diff
@@ -209,10 +209,11 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
+    # Retrieve name of AMI, if ClientError or AttributeError is thrown, the AMI does not exist
+    try:
+        ec2.Image(ami_id).name
+    except (botocore.exceptions.ClientError, AttributeError):
```
`AttributeError` seems suspicious...

alizbay94 replied:
I agree - I don't know if you saw my previous comment, but I discovered a bug with an open ticket [here](https://github.com/boto/boto3/issues/925) on the boto3 project’s GitHub page describing how None type is returned for deregistered amis, and the error that is thrown is: `AttributeError: 'NoneType' object has no attribute 'get'`.
If its too suspicious I can always go back to the other method of getting a list of our AMIs and checking them. I'm alright with using the `AttributeError` with the `ClientError` because I can't anticipate serious ramifications, but I may be overlooking something.

CodyRay commented on app/ami.py:
```diff
@@ -209,10 +209,11 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
 
 
 def is_ami_registered(ami_id: str, ) -> bool:
```
Seems to be an extra `,`
```suggestion
def is_ami_registered(ami_id: str) -> bool:
```

CodyRay commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
 boto3==1.20.26
+botocore==1.23.54
```
Should these two be the same version?

alizbay94 replied:
They should work as different versions, I looked online and saw the [latest version of boto3](https://github.com/boto/boto3/tags) is 1.25.3 and [the latest version of botocore](https://github.com/boto/botocore/tags) is 1.28.3. I imagine if they had to be the same version the latest versions would be the same, and there weren't any conflicts on my machine when I ran nagbot (I have those two versions running).

alizbay94 committed:
remove extra string from method params

Co-authored-by: Cody Ray Hoeft <cody.ray.hoeft@seeq.com>

CodyRay commented:
> > I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)
> 
> Sounds good i'll implement that so they are all deleted as one happy family.

Wanted to give this a poke because I think it is the only piece of feedback from me AFAIK

alizbay94 committed:
Delete Snapshots making up EBS-backed AMI when deregistering AMI

alizbay94 committed:
Merge branch 'task/azb/support-ebs-snapshot-and-amis-CRAB-33767' of https://github.com/seeq12/nagbot into task/azb/support-ebs-snapshot-and-amis-CRAB-33767

alizbay94 commented:
Ended up extracting the snapshot ids from the block device mappings of AMIs, then created aws ec2.Snapshot objects using those extracted ids. After the list of snapshot objects is created, it is assigned to the "snapshots" attribute of the Ami object that is created. I added a "snapshots" attribute to the Ami class that represents a list of aws ec2.Snapshot objects that make up the Ami. During the method when an Ami is terminated (deregistered), each of the Snapshots belonging to the Ami are also deleted.

alizbay94 committed:
Return list of id strings instead of snapshot objects, and incorporate Pytests for Snapshot deletion of AMIs after AMI deregistering

alizbay94 committed:
Increment version of NagBot to v1.11.0
